### PR TITLE
Patch: r2.5.0 with onelogger changes.

### PR DESCRIPTION
**Pull request by PeiyuanQi**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Patch release r2.5.0 with changes related to nv-one-logger which enable NeMo using OneLogger for job performance instrumentation.

Original PR [!13437](https://github.com/NVIDIA-NeMo/NeMo/pull/13437)

**Collection**: [Note which collection this PR will affect]
Will keep updating
- llm
- speechlm2
- vlm
- core
- lightning

# Changelog

- Add new `CallbackGroup` and `Callback` ABC in `NeMo/nemo/lightning/pytorch/callbacks/callback_group.py` and `base_callback.py` used by various collections.
- `collections/llm`
  - Add `try` `finally` logic for trainer in `collections/llm/api.py` to ensure the `on_app_end()` call is always called.
  - Add `__init_subclass__` method to ensure the callbacks are hooked in `collections/llm/fn/mixin.py`.
  - Fix minor fail safe check in `collections/llm/modelopt/speculative/model_transform.py`.
  -  Add data mocks for `collections/llm/gpt/data/mock.py`, `collections/llm/t5/data/mock.py`.
- `collections/speechlm2`
  - minor fix of regex in `optim_setup.py`.
- `collections/vlm`
  - add `on_dataloader_init_start/end()` instrumentation hook call in `preloaded.py`.
- `collections/core`
  - add `one_model_init_start/end()` `on_dataloader_init_start/end()`, `on_load_checkpoint_start/end()` instrumentation hook call in `modelPT.py`.
  - fix minor fail safe code value of `optimizer_name` in `modelPT.py`.
  - add `finally` to make sure with exception in `_run_hydra` still instrument `on_app_end()`.
  - fix circular dependency issue in `optim/lr_scheduler.py`
  - fix circular dependency issue in `optim/optimizers.py`
- `collections/lightning`
  - add `callback` related classes.
  - import hook of `OneLogger` module for lightning boot up in `collections/lightning/io/mixin.py`
  - fix bug in `model_checkpoint.py` which result in pipeline failure in `collections/lightning/pytorch/callbacks`.
- `tests`
  - related unit tests are added.

# Usage

Not changed.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by chtruong814** [comment]
Import error is unrelated to this change.  Going forward with merging this.
